RIOT-OS/RIOT
# cpu/esp32: improvements and cleanup of log_module

### Contribution description

This PR is an improvement of the `log_module` for ESP32. It consists of the following changes:

- The `log_module` for ESP32 now supports colored log outputs. Colored log outputs are activated when the `esp_log_color` module is used. The generation of colored log outputs is realized by extending the bunch of the existing log generation macros with the type of the log message.

- For the implementation of the colored log output, two different versions of the bootloader are used, one version with colored log output and one version without colors. The bootloader version with colored log outputs is used when the  `esp_log_color` module is used. Otherwise, the bootloader version without colored log outputs is used.

- ESP32 log outputs were tagged with additional information by default. These additional information  includes the type of the log message, the system time in ms, and the module or function in which the log message is generated. By introducing module `esp_log_tagged`, these additional information are disabled by default and can be enabled explicitly by using module `esp_log_tagged`.

- The implementation of the `log_module` for ESP32 is changed from the former function-based implementation to a macro-based implementation to be able to use the bunch of macros for colored and tagged log output generation.

- Log outputs generated by binary ESP32 SDK libraries are now mapped to the `log_module` which supports colored and tagged log outpus. Tagged log outputs from SDK libraries are handled accordingly.

- Wherever possible, `ets_printf` is replaced by the newlib `printf`.

[UPDATE] With PR #11573, the system module `log_color` was introduced. If module `log_color` is enabled, module `esp_log_colored` is enabled automatically. The test application `tests/log_color` is working with this PR.

### Testing procedure

1. Flash and test `examples/hello-world` using command:
   ```
   make BOARD=esp32-wroom-32 -C examples/hello-world
   ```
   The output should not contain additional information and should not be colored.
   ```
   main(): This is RIOT! (Version: 2019.10-devel-437-gf506a-cpu/esp32/log_module_fix)
   Hello World!
   ```
   Note: The bootloader messages are always tagged with the additional information.

2. Flash and test `examples/hello-world` with tagged log output using command:
   ```
   USEMODULE=esp_log_tagged make BOARD=esp32-wroom-32 -C examples/hello-world flash
   ```
   The output should contain `I` for info message, `(...)` the system time and `[main_trampoline]` the function were it is generated:
   ```
   I (663) [main_trampoline] main(): This is RIOT! (Version: 2019.10-devel-437-gf506a-cpu/esp32/log_module_fix)
   Hello World!
   ```
3. Flash and test `examples/hello-world` with tagged log and colored output using command:
   ```
   USEMODULE='esp_log_tagged esp_log_colored' make BOARD=esp32-wroom-32 -C examples/hello-world flash
   ```
   Use a terminal program which is able to interpret ANSI escape sequences for colors, for example:
   ```
   picocom -b 115200 --imap lfcrlf /dev/ttyUSB0
   ```
   The log output of application and the bootloader should be colored.

4. Flash and test the more complex application `examples/gnrc_networking` with tagged log and colored output to observe a more complex tagged and colored log output:
   ```
   USEMODULE='esp_wifi esp_log_tagged esp_log_colored' \
   CFLAGS='-DESP_WIFI_SSID=\"ssid\" -DESP_WIFI_PASS=\"passphrase\"' \
   make BOARD=esp32-wroom-32 -C examples/gnrc_networking flash
   ```

### Issues/PRs references

Fixes also the problem with `tests/log_printfnoformat` in PR #11449.
